packages/feedback-components/src/feedback/text-visualizer.ts

```diff
@@ -78,95 +82,199 @@ export function FeedbackText(props: FeedbackTextProps) {
+  dispatch({ type: "create-node", payload });
+}
 
-        if (removedIds.length > 0) {
-          dispatch({
-            type: "toggle-node-selected",
-            payload: { ids: removedIds },
-          });
+export function HighlightedText(props: {
+  text: string;
+  allTags: AnnotateBlendTag[];
+  lineHeight?: string;
+  dispatch: any;
+  allowOverlap: boolean;
+  selectedNodes: any[];
+}) {
+  const { text, allTags = [], lineHeight, dispatch, allowOverlap, selectedNodes } = props;
+  const parts = [];
+  let start = 0;
+  const spanRef = useRef<HTMLSpanElement>(null);
+
+  useEffect(() => {
+    const handleMouseUp = () => {
+      const tag = createTagFromSelection({ container: spanRef.current });
+      if (!tag) return;
+      addTag({ tag, dispatch, text, allTags, allowOverlap });
+    };
+
+    document.addEventListener("mouseup", handleMouseUp);
+    return () => {
+      document.removeEventListener("mouseup", handleMouseUp);
+    };
+  }, [text, allTags, dispatch, allowOverlap]);
+
+  const sortedTags = [...allTags].sort((a, b) => {
+    if (a.start === b.start) {
+      return (a.end - a.start) - (b.end - b.start);
+    }
+    return a.start - b.start;
+  });
+
+  const selectedTag = allTags.find(tag => selectedNodes?.includes(tag.id));
+  const claimedRanges: Array<[number, number]> = [];
+
+  function subtractOverlaps(start: number, end: number): Array<[number, number]> {
+    let ranges: Array<[number, number]> = [[start, end]];
+
+    for (const [cStart, cEnd] of claimedRanges) {
+      const newRanges: typeof ranges = [];
+      for (const [rStart, rEnd] of ranges) {
+        if (cEnd <= rStart || cStart >= rEnd) {
+          newRanges.push([rStart, rEnd]);
+        } else {
+          if (rStart < cStart) newRanges.push([rStart, cStart]);
+          if (rEnd > cEnd) newRanges.push([cEnd, rEnd]);
         }
       }
-    },
-    [allTags, text]
-  );
+      ranges = newRanges;
+      if (ranges.length === 0) break;
+    }
 
-  const value = allTags
-  console.log("FeedbackText value", value);
+    return ranges;
+  }
 
-  return h('div.feedback-text-wrapper', { 
-    tabIndex: 0,
-    onKeyDown: (e) => {
-      if( e.key === "Backspace") {
-        dispatch({
-          type: "delete-node",
-          payload: { ids: selectedNodes },
-        });
+  for (const tag of sortedTags) {
+    // Skip inner tags when a larger tag is selected
+    if (
+      selectedTag &&
+      selectedTag.id !== tag.id &&
+      tag.start >= selectedTag.start &&
+      tag.end <= selectedTag.end
+    ) {
+      continue;
+    }
+
+    const { start: tagStart, end: tagEnd, ...rest } = tag;
+    const availableRanges = subtractOverlaps(tagStart, tagEnd);
+
+    for (const [s, e] of availableRanges) {
+      if (start < s) {
+        parts.push(text.slice(start, s));
       }
+
+      parts.push(
+        h(
+          "span.highlight",
+          {
+            style: {
+              ...rest,
+              position: "relative",
```

Comment:
Try to do some of this basic styling in SCSS rather than manipulating styles in JS